enclave/core/sgx/once.c

```diff
@@ -5,34 +5,39 @@
 #include <openenclave/internal/thread.h>
 #include <openenclave/internal/utils.h>
 
+#define SPINLOCK_NOT_USED 0
+#define SPINLOCK_ACTIVELY_HELD 1
+#define SPINLOCK_USED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
     OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
-    if (o == 0)
+    if (*once != SPINLOCK_USED)
     {
-        static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
-
-        oe_spin_lock(&_lock);
-
-        if (*once == 0)
+        oe_once_t retval = __sync_val_compare_and_swap(
+            once, SPINLOCK_NOT_USED, SPINLOCK_ACTIVELY_HELD);
+        if (retval == SPINLOCK_NOT_USED)
         {
             if (func)
                 func();
 
-            /* DCLP Release barrier. */
-            OE_ATOMIC_MEMORY_BARRIER_RELEASE();
-            *once = 1;
+            *once = SPINLOCK_USED;
```

Comment:
This line ought to be preceded by 
OE_ATOMIC_MEMORY_BARRIER_RELEASE();

The concept to acquire-release is this:
Say, func all func does is set g to some value;

void func()
{
    g = MAGIC;
}

Eventhough, as written, the order is:
    g = MAGIC;
   *once = SPINLOCK_USED; // or FUNC_INVOKED
the compiler, could reorder this to:
   *once = SPINLOCK_USED;
    g = MAGIC
Even if compiler preserves the order, another thread could see it in the incorrect order when contents of one processor core is flushed and synced to another core. That is another core could see these variables being written out of order.
So, these ACQUIRE/RELESE macros insert special memory barriers so that different cores in a multi-core system see the writes in the correct order.

Note: x64 is strongly ordered, and these marcros are no-ops. But other processors are not.

Reply:
I just made your suggested change for this, but I'm going to leave this comment unresolved for the really great & important information you left here about concerns regarding ordering of operations in this scenario.